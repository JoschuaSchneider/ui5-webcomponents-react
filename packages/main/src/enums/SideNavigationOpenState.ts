export enum SideNavigationOpenState {
  Expanded = 'Expanded',
  Condensed = 'Condensed'
}
