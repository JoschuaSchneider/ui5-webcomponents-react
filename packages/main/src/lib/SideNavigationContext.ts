import { SideNavigationOpenState } from '@ui5/webcomponents-react/lib/SideNavigationOpenState';
import { createContext } from 'react';

interface ISideNavigationContext {
  openState: SideNavigationOpenState;
  selectedId: string;
  noIcons: boolean;
  onListItemSelected: () => void;
  expandedItems: string[];
  toggleExpandedItems: (id: string) => void;
}

const SideNavigationContext = createContext<ISideNavigationContext>({
  openState: SideNavigationOpenState.Expanded,
  selectedId: null,
  noIcons: false,
  onListItemSelected: () => {},
  toggleExpandedItems: () => {},
  expandedItems: []
});

export { ISideNavigationContext, SideNavigationContext };
