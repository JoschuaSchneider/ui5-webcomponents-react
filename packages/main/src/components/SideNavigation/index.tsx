import { createComponentStyles } from '@ui5/webcomponents-react-base/lib/createComponentStyles';
import { useConsolidatedRef, usePassThroughHtmlProps } from '@ui5/webcomponents-react-base/lib/hooks';
import { StyleClassHelper } from '@ui5/webcomponents-react-base/lib/StyleClassHelper';
import { enrichEventWithDetails } from '@ui5/webcomponents-react-base/lib/Utils';
import { List } from '@ui5/webcomponents-react/lib/List';
import { SideNavigationContext } from '@ui5/webcomponents-react/lib/SideNavigationContext';
import { SideNavigationOpenState } from '@ui5/webcomponents-react/lib/SideNavigationOpenState';
import React, {
  Children,
  cloneElement,
  FC,
  forwardRef,
  ReactNode,
  Ref,
  useCallback,
  useEffect,
  useRef,
  useState
} from 'react';
import { CommonProps } from '../../interfaces/CommonProps';
import { sideNavigationStyles } from './SideNavigation.jss';

export interface SideNavigationProps extends CommonProps {
  openState?: SideNavigationOpenState;
  children?: ReactNode;
  footerItems?: ReactNode[];
  selectedId?: string | number;
  onItemSelect?: (event: CustomEvent<{ selectedItem: HTMLElement; selectedId: string | number }>) => void;
  onItemClick?: (
    event: CustomEvent<{
      selectedItem: HTMLElement;
      selectedId: string | number;
      hasChildren: boolean;
      isExpanded: boolean;
    }>
  ) => void;
  /*
   * Flag whether to show icons or not. Will only take effect in <code>openState: Expanded</code>
   */
  noIcons?: boolean;
}

let lastFiredSelection = '';

const useStyles = createComponentStyles(sideNavigationStyles, {
  name: 'SideNavigation'
});

/**
 * <code>import { SideNavigation } from '@ui5/webcomponents-react/lib/SideNavigation';</code>
 */
const SideNavigation: FC<SideNavigationProps> = forwardRef((props: SideNavigationProps, ref: Ref<HTMLDivElement>) => {
  const {
    children,
    openState,
    footerItems,
    selectedId,
    onItemSelect,
    onItemClick,
    noIcons,
    style,
    className,
    tooltip
  } = props;

  const classes = useStyles();

  const sideNavigationRef = useConsolidatedRef<any>(ref);

  const [internalSelectedId, setInternalSelectedId] = useState(selectedId);
  const [expandedItems, setExpandedItems] = useState(() => Children.map(children, (child) => child.props.id));
  const selectionTimeout = useRef(null);

  useEffect(() => {
    sideNavigationRef.current.expandItem = (id) => {
      setExpandedItems((items) => Array.from(new Set([items, id])));
    };
    sideNavigationRef.current.collapseItem = (id) => {
      setExpandedItems((items) => items.filter((item) => item !== id));
    };

    sideNavigationRef.current.toggleItem = (id) => {
      setExpandedItems((items) => {
        if (items.includes(id)) {
          return items.filter((item) => item !== id);
        }
        return Array.from(new Set([items, id]));
      });
    };
  }, [sideNavigationRef.current, setExpandedItems]);

  useEffect(() => {
    setInternalSelectedId(selectedId);
  }, [selectedId, setInternalSelectedId]);

  const sideNavigationClasses = StyleClassHelper.of(classes.sideNavigation).putIfPresent(className);

  switch (openState) {
    case SideNavigationOpenState.Expanded: {
      sideNavigationClasses.put(classes.expanded);
      break;
    }
    case SideNavigationOpenState.Condensed: {
      sideNavigationClasses.put(classes.condensed);
      break;
    }
  }

  const toggleExpandedItems = useCallback(
    (id) => {
      cancelAnimationFrame(selectionTimeout.current);
      setExpandedItems((items) => {
        if (items.includes(id)) {
          return items.filter((item) => item !== id);
        }
        return Array.from(new Set([...items, id]));
      });
    },
    [setExpandedItems]
  );

  const onListItemSelected = useCallback(
    (e) => {
      selectionTimeout.current = requestAnimationFrame(() => {
        const listItem = e.detail.item;
        onItemClick(
          enrichEventWithDetails(e, {
            selectedItem: listItem,
            selectedId: listItem.dataset.id,
            hasChildren: listItem.hasAttribute('data-has-children'),
            isExpanded: listItem.hasAttribute('data-is-expanded')
          })
        );

        if (lastFiredSelection === listItem.dataset.id) {
          return;
        }
        setInternalSelectedId(listItem.dataset.id);

        onItemSelect(
          enrichEventWithDetails(e, {
            selectedItem: listItem,
            selectedId: listItem.dataset.id
          })
        );
        lastFiredSelection = listItem.dataset.id;
        selectionTimeout.current = null;
      });
    },
    [onItemSelect, onItemClick, setInternalSelectedId, setExpandedItems, toggleExpandedItems, selectionTimeout]
  );

  const passThroughProps = usePassThroughHtmlProps(props, ['onItemSelect', 'onItemClick']);

  const contextValue = {
    openState,
    selectedId: internalSelectedId,
    noIcons,
    onListItemSelected,
    expandedItems: expandedItems,
    toggleExpandedItems
  };

  return (
    <div
      ref={sideNavigationRef}
      className={sideNavigationClasses.className}
      style={style}
      title={tooltip}
      {...passThroughProps}
    >
      <SideNavigationContext.Provider value={contextValue}>
        <List onItemClick={onListItemSelected}>{children}</List>
        <span style={{ flexGrow: 1 }} />
        {footerItems.length > 0 && <span className={classes.footerItemsSeparator} />}
        {footerItems && (
          <List onItemClick={onListItemSelected}>
            {footerItems.map((item: any, index) =>
              cloneElement(item, {
                key: index
              })
            )}
          </List>
        )}
      </SideNavigationContext.Provider>
    </div>
  );
});

SideNavigation.displayName = 'SideNavigation';

SideNavigation.defaultProps = {
  openState: SideNavigationOpenState.Expanded,
  footerItems: [],
  selectedId: null,
  onItemClick: () => {},
  onItemSelect: () => {}
};

export { SideNavigation };
