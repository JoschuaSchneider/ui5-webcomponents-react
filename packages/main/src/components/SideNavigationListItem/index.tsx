import '@ui5/webcomponents-icons/dist/icons/navigation-down-arrow.js';
import '@ui5/webcomponents-icons/dist/icons/navigation-right-arrow.js';
import { createComponentStyles } from '@ui5/webcomponents-react-base/lib/createComponentStyles';
import { StyleClassHelper } from '@ui5/webcomponents-react-base/lib/StyleClassHelper';
import { usePassThroughHtmlProps } from '@ui5/webcomponents-react-base/lib/usePassThroughHtmlProps';
import { CustomListItem } from '@ui5/webcomponents-react/lib/CustomListItem';
import { Icon } from '@ui5/webcomponents-react/lib/Icon';
import { List } from '@ui5/webcomponents-react/lib/List';
import { Popover } from '@ui5/webcomponents-react/lib/Popover';
import { PopoverVerticalAlign } from '@ui5/webcomponents-react/lib/PopoverVerticalAlign';
import { ISideNavigationContext, SideNavigationContext } from '@ui5/webcomponents-react/lib/SideNavigationContext';
import { SideNavigationOpenState } from '@ui5/webcomponents-react/lib/SideNavigationOpenState';
import { StandardListItem } from '@ui5/webcomponents-react/lib/StandardListItem';
import { Text } from '@ui5/webcomponents-react/lib/Text';

import React, {
  Children,
  cloneElement,
  FC,
  forwardRef,
  ReactNode,
  ReactNodeArray,
  Ref,
  useCallback,
  useContext,
  useMemo,
  useRef
} from 'react';
import { createPortal } from 'react-dom';
import { CommonProps } from '../../interfaces/CommonProps';
import { sideNavigationListItemStyles } from './SideNavigationListItem.jss';

export interface SideNavigationListItemProps extends CommonProps {
  icon?: string;
  text: string;
  id: number | string;
  children?: ReactNode | ReactNodeArray;
  expanded?: boolean;
}

const useStyles = createComponentStyles(sideNavigationListItemStyles, {
  name: 'SideNavigationListItem'
});
/**
 * <code>import { SideNavigationListItem } from '@ui5/webcomponents-react/lib/SideNavigationListItem';</code>
 */
const SideNavigationListItem: FC<SideNavigationListItemProps> = forwardRef(
  (props: SideNavigationListItemProps, ref: Ref<HTMLDivElement>) => {
    const { icon, text, id, children, tooltip, slot, className, style } = props;

    const context = useContext<ISideNavigationContext>(SideNavigationContext);

    const isExpanded = context.expandedItems.includes(id);
    const handleToggleExpand = useCallback(
      (e) => {
        e.preventDefault();
        e.stopPropagation();
        context.toggleExpandedItems(id);
      },
      [context.toggleExpandedItems, id]
    );

    const classes = useStyles();

    const listItemClasses = StyleClassHelper.of(classes.listItem);

    if (className) {
      listItemClasses.put(className);
    }

    if (context.noIcons) {
      listItemClasses.put(classes.noIcons);
    }

    const validChildren = Children.toArray(children).filter(Boolean);
    const hasChildren = validChildren.length > 0;

    const isOpenStateExpanded = context.openState === SideNavigationOpenState.Expanded;

    const isSelfSelected = context.selectedId === id;
    const hasSelectedChild = !!validChildren.find((child: any) => child.props.id === context.selectedId);

    const passThroughProps = usePassThroughHtmlProps(props, ['id']);

    const customListItemCommonProps = {
      ref,
      className: listItemClasses.valueOf(),
      tooltip,
      slot,
      style,
      'data-id': id,
      ...passThroughProps
    };

    if (isExpanded && isOpenStateExpanded && hasChildren) {
      customListItemCommonProps['data-is-expanded'] = '';
    }

    if (hasChildren) {
      customListItemCommonProps['data-has-children'] = '';
    }

    if (props['isChild']) {
      customListItemCommonProps['data-is-child'] = '';
    }

    const popoverRef = useRef();

    const displayedIcon = useMemo(() => {
      if (!icon) {
        return <span className={classes.icon} />;
      }
      return <Icon name={icon} className={classes.icon} />;
    }, [classes.icon, icon]);

    const handleOpenPopover = useCallback(
      (e) => {
        // @ts-ignore
        popoverRef.current.openBy(e.target);
      },
      [popoverRef.current]
    );

    if (isOpenStateExpanded) {
      return (
        <>
          <CustomListItem selected={isSelfSelected} {...customListItemCommonProps}>
            {!context.noIcons && displayedIcon}
            <Text className={classes.text}>{text}</Text>
            {hasChildren && (
              <Icon
                name={isExpanded ? 'navigation-down-arrow' : 'navigation-right-arrow'}
                onClick={handleToggleExpand}
                className={classes.expandArrow}
              />
            )}
          </CustomListItem>
          {isExpanded &&
            validChildren.map((child: any, index: number) => {
              const style = child.props.style || {};
              if (index !== validChildren.length - 1) {
                style['--ui5-listitem-border-bottom'] = 'none';
              }

              return cloneElement(child, {
                icon: null,
                style,
                isChild: true
              });
            })}
        </>
      );
    }

    return (
      <CustomListItem selected={isSelfSelected || hasSelectedChild} {...customListItemCommonProps}>
        {hasChildren ? (
          <span onClick={handleOpenPopover}>
            {displayedIcon}
            <div className={classes.condensedExpandTriangle} />
            {createPortal(
              <Popover ref={popoverRef} verticalAlign={PopoverVerticalAlign.Top}>
                <List onItemClick={context.onListItemSelected}>
                  <StandardListItem selected={isSelfSelected} data-id={id} tooltip={tooltip}>
                    {text}
                  </StandardListItem>
                  {validChildren.map((child: any, index) => (
                    <StandardListItem
                      selected={context.selectedId === child.props.id}
                      key={index}
                      data-id={child.props.id}
                      tooltip={child.props.tooltip || child.props.text}
                    >
                      {child.props.text}
                    </StandardListItem>
                  ))}
                </List>
              </Popover>,
              document.body
            )}
          </span>
        ) : (
          displayedIcon
        )}
      </CustomListItem>
    );
  }
);

SideNavigationListItem.displayName = 'SideNavigationListItem';

SideNavigationListItem.defaultProps = {};

export { SideNavigationListItem };
