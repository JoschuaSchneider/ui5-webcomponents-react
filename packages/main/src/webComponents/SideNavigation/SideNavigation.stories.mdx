import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks';
import { SideNavigation } from '@ui5/webcomponents-react/lib/SideNavigation';

import { SideNavigationItem } from '@ui5/webcomponents-react/lib/SideNavigationItem';
import { SideNavigationSubItem } from '@ui5/webcomponents-react/lib/SideNavigationSubItem';
import { createSelectArgTypes } from '@shared/stories/createSelectArgTypes';
import { DocsHeader } from '@shared/stories/DocsHeader';

<Meta
  title="UI5 Web Components / SideNavigation"
  component={SideNavigation}
  subcomponents={{ SideNavigationItem, SideNavigationSubItem }}
  argTypes={{
    ...createSelectArgTypes({})
  }}
  args={{}}
/>

<DocsHeader />

<Canvas>
  <Story name="Default">
    {(args) => {
      return <SideNavigation {...args} />;
    }}
  </Story>
</Canvas>

<ArgsTable story="." />
<h3>Usage</h3>

Use the available <code>SideNavigationItem</code> and <code>ui5-side-navigation-sub-item</code> components to build your
menu. The items can consist of text only or an icon with text. The use or non-use of icons must be consistent for all
items on one level. You must not combine entries with and without icons on the same level. We strongly recommend that
you do not use icons on the second level.
