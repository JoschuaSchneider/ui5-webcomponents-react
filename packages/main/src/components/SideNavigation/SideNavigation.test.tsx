import { createPassThroughPropsTest } from '@shared/tests/utils';
import { act, fireEvent, render, screen, cleanup } from '@tests/index';
import { SideNavigation } from '@ui5/webcomponents-react/lib/SideNavigation';
import { SideNavigationListItem } from '@ui5/webcomponents-react/lib/SideNavigationListItem';
import { SideNavigationOpenState } from '@ui5/webcomponents-react/lib/SideNavigationOpenState';
import { mount } from 'enzyme';
import React, { createRef } from 'react';
import { SideNavigationRef } from './index';

const sideNavigationContent = (
  <>
    <SideNavigationListItem text="Overview" icon="home" id="home" tooltip="Nav To Home Tooltip" />
    <SideNavigationListItem text="Calendar" icon="calendar" id="calendar" />
    <SideNavigationListItem text="Customers" icon="employee" id="customers" />
    <SideNavigationListItem text="Sales" icon="lead" id="sales">
      <SideNavigationListItem text="My Opportunities" icon="home" id="sales-opportunities" />
      <SideNavigationListItem text="My Leads" icon="home" id="sales-leads" />
      <SideNavigationListItem text="My CPQS" icon="home" id="sales-cpqs" />
    </SideNavigationListItem>
    <SideNavigationListItem text="Deliveries" icon="add-product" id="delivieries" />
  </>
);

describe('SideNavigation', () => {
  afterEach(cleanup);

  test('expand and collapse via ref', async () => {
    const ref = createRef<SideNavigationRef>();
    render(<SideNavigation ref={ref}>{sideNavigationContent}</SideNavigation>);

    // all items should be expanded initially
    expect(screen.getByText('My Leads')).toBeInTheDocument();

    // now collapse sales
    act(() => {
      ref.current.collapseItem('sales');
    });
    expect(screen.queryByText('My Leads')).not.toBeInTheDocument();

    // expand it again
    act(() => {
      ref.current.expandItem('sales');
    });
    expect(screen.getByText('My Leads')).toBeInTheDocument();

    // toggle
    act(() => {
      ref.current.toggleItem('sales');
    });
    expect(screen.queryByText('My Leads')).not.toBeInTheDocument();
    act(() => {
      ref.current.toggleItem('sales');
    });
    expect(screen.getByText('My Leads')).toBeInTheDocument();
  });

  test('Expanded', () => {
    const wrapper = mount(
      <SideNavigation
        openState={SideNavigationOpenState.Expanded}
        selectedId={'sales-leads'}
        footerItems={[
          <SideNavigationListItem id="1" text="Legal Information" icon="compare" />,
          <SideNavigationListItem id="2" text="Useful Links" icon="chain-link" />
        ]}
      >
        {sideNavigationContent}
        <SideNavigationListItem text="Support Tickets" icon="add-product" id="support" />
      </SideNavigation>
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  test('Expanded without Icons', () => {
    const wrapper = mount(
      <SideNavigation
        openState={SideNavigationOpenState.Expanded}
        selectedId={'sales-leads'}
        noIcons
        footerItems={[
          <SideNavigationListItem id="1" text="Legal Information" icon="compare" />,
          <SideNavigationListItem id="2" text="Useful Links" icon="chain-link" />
        ]}
      >
        {sideNavigationContent}
      </SideNavigation>
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  test('Condensed', () => {
    const wrapper = mount(
      <SideNavigation
        openState={SideNavigationOpenState.Condensed}
        selectedId={'sales-leads'}
        footerItems={[
          <SideNavigationListItem id="1" text="Legal Information" icon="compare" />,
          <SideNavigationListItem id="2" text="Useful Links" icon="chain-link" />
        ]}
      >
        {sideNavigationContent}
      </SideNavigation>
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  test.skip('item selection', async () => {
    const onListItemSelect = jest.fn();
    const onListItemClick = jest.fn();

    render(
      <SideNavigation selectedId="sales-leads" onItemClick={onListItemClick} onItemSelect={onListItemSelect}>
        {sideNavigationContent}
      </SideNavigation>
    );

    const customers = screen.getByText('Customers');
    fireEvent.click(customers.parentElement.shadowRoot.querySelector('li'));

    // screen.debug();

    expect(onListItemClick).toBeCalled();
    expect(onListItemSelect).toBeCalled();

    fireEvent.click(screen.getByText('Customers'));
    expect(onListItemClick).toBeCalledTimes(2);
    expect(onListItemSelect).toBeCalledTimes(1);
  });

  createPassThroughPropsTest(SideNavigation, { children: sideNavigationContent });
});
