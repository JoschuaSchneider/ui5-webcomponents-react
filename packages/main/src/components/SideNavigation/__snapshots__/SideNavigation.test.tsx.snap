// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideNavigation Condensed 1`] = `
<div
  class="SideNavigation-sideNavigation-0 SideNavigation-condensed-0"
>
  <ui5-list
    mode="None"
    separators="All"
  >
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0"
      data-id="home"
      id="home"
      title="Nav To Home Tooltip"
      type="Active"
    >
      <ui5-icon
        class="SideNavigationListItem-icon-0"
        name="home"
      />
    </ui5-li-custom>
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0"
      data-id="calendar"
      id="calendar"
      type="Active"
    >
      <ui5-icon
        class="SideNavigationListItem-icon-0"
        name="calendar"
      />
    </ui5-li-custom>
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0"
      data-id="customers"
      id="customers"
      type="Active"
    >
      <ui5-icon
        class="SideNavigationListItem-icon-0"
        name="employee"
      />
    </ui5-li-custom>
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0"
      data-has-children=""
      data-id="sales"
      id="sales"
      selected="true"
      type="Active"
    >
      <span>
        <ui5-icon
          class="SideNavigationListItem-icon-0"
          name="lead"
        />
        <div
          class="SideNavigationListItem-condensedExpandTriangle-0"
        />
      </span>
    </ui5-li-custom>
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0"
      data-id="delivieries"
      id="delivieries"
      type="Active"
    >
      <ui5-icon
        class="SideNavigationListItem-icon-0"
        name="add-product"
      />
    </ui5-li-custom>
  </ui5-list>
  <span
    style="flex-grow: 1;"
  />
  <span
    class="SideNavigation-footerItemsSeparator-0"
  />
  <ui5-list
    mode="None"
    separators="All"
  >
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0"
      data-id="1"
      id="1"
      type="Active"
    >
      <ui5-icon
        class="SideNavigationListItem-icon-0"
        name="compare"
      />
    </ui5-li-custom>
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0"
      data-id="2"
      id="2"
      type="Active"
    >
      <ui5-icon
        class="SideNavigationListItem-icon-0"
        name="chain-link"
      />
    </ui5-li-custom>
  </ui5-list>
</div>
`;

exports[`SideNavigation Expanded 1`] = `
<div
  class="SideNavigation-sideNavigation-0 SideNavigation-expanded-0"
>
  <ui5-list
    mode="None"
    separators="All"
  >
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0"
      data-id="home"
      id="home"
      title="Nav To Home Tooltip"
      type="Active"
    >
      <ui5-icon
        class="SideNavigationListItem-icon-0"
        name="home"
      />
      <span
        class="Text-text-0 SideNavigationListItem-text-0"
      >
        Overview
      </span>
    </ui5-li-custom>
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0"
      data-id="calendar"
      id="calendar"
      type="Active"
    >
      <ui5-icon
        class="SideNavigationListItem-icon-0"
        name="calendar"
      />
      <span
        class="Text-text-0 SideNavigationListItem-text-0"
      >
        Calendar
      </span>
    </ui5-li-custom>
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0"
      data-id="customers"
      id="customers"
      type="Active"
    >
      <ui5-icon
        class="SideNavigationListItem-icon-0"
        name="employee"
      />
      <span
        class="Text-text-0 SideNavigationListItem-text-0"
      >
        Customers
      </span>
    </ui5-li-custom>
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0"
      data-has-children=""
      data-id="sales"
      data-is-expanded=""
      id="sales"
      type="Active"
    >
      <ui5-icon
        class="SideNavigationListItem-icon-0"
        name="lead"
      />
      <span
        class="Text-text-0 SideNavigationListItem-text-0"
      >
        Sales
      </span>
      <ui5-icon
        class="SideNavigationListItem-expandArrow-0"
        name="navigation-down-arrow"
      />
    </ui5-li-custom>
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0"
      data-id="sales-opportunities"
      data-is-child=""
      id="sales-opportunities"
      style="--ui5-listitem-border-bottom: none;"
      type="Active"
    >
      <span
        class="SideNavigationListItem-icon-0"
      />
      <span
        class="Text-text-0 SideNavigationListItem-text-0"
      >
        My Opportunities
      </span>
    </ui5-li-custom>
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0"
      data-id="sales-leads"
      data-is-child=""
      id="sales-leads"
      selected="true"
      style="--ui5-listitem-border-bottom: none;"
      type="Active"
    >
      <span
        class="SideNavigationListItem-icon-0"
      />
      <span
        class="Text-text-0 SideNavigationListItem-text-0"
      >
        My Leads
      </span>
    </ui5-li-custom>
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0"
      data-id="sales-cpqs"
      data-is-child=""
      id="sales-cpqs"
      type="Active"
    >
      <span
        class="SideNavigationListItem-icon-0"
      />
      <span
        class="Text-text-0 SideNavigationListItem-text-0"
      >
        My CPQS
      </span>
    </ui5-li-custom>
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0"
      data-id="delivieries"
      id="delivieries"
      type="Active"
    >
      <ui5-icon
        class="SideNavigationListItem-icon-0"
        name="add-product"
      />
      <span
        class="Text-text-0 SideNavigationListItem-text-0"
      >
        Deliveries
      </span>
    </ui5-li-custom>
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0"
      data-id="support"
      id="support"
      type="Active"
    >
      <ui5-icon
        class="SideNavigationListItem-icon-0"
        name="add-product"
      />
      <span
        class="Text-text-0 SideNavigationListItem-text-0"
      >
        Support Tickets
      </span>
    </ui5-li-custom>
  </ui5-list>
  <span
    style="flex-grow: 1;"
  />
  <span
    class="SideNavigation-footerItemsSeparator-0"
  />
  <ui5-list
    mode="None"
    separators="All"
  >
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0"
      data-id="1"
      id="1"
      type="Active"
    >
      <ui5-icon
        class="SideNavigationListItem-icon-0"
        name="compare"
      />
      <span
        class="Text-text-0 SideNavigationListItem-text-0"
      >
        Legal Information
      </span>
    </ui5-li-custom>
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0"
      data-id="2"
      id="2"
      type="Active"
    >
      <ui5-icon
        class="SideNavigationListItem-icon-0"
        name="chain-link"
      />
      <span
        class="Text-text-0 SideNavigationListItem-text-0"
      >
        Useful Links
      </span>
    </ui5-li-custom>
  </ui5-list>
</div>
`;

exports[`SideNavigation Expanded without Icons 1`] = `
<div
  class="SideNavigation-sideNavigation-0 SideNavigation-expanded-0"
>
  <ui5-list
    mode="None"
    separators="All"
  >
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0 SideNavigationListItem-noIcons-0"
      data-id="home"
      id="home"
      title="Nav To Home Tooltip"
      type="Active"
    >
      <span
        class="Text-text-0 SideNavigationListItem-text-0"
      >
        Overview
      </span>
    </ui5-li-custom>
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0 SideNavigationListItem-noIcons-0"
      data-id="calendar"
      id="calendar"
      type="Active"
    >
      <span
        class="Text-text-0 SideNavigationListItem-text-0"
      >
        Calendar
      </span>
    </ui5-li-custom>
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0 SideNavigationListItem-noIcons-0"
      data-id="customers"
      id="customers"
      type="Active"
    >
      <span
        class="Text-text-0 SideNavigationListItem-text-0"
      >
        Customers
      </span>
    </ui5-li-custom>
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0 SideNavigationListItem-noIcons-0"
      data-has-children=""
      data-id="sales"
      data-is-expanded=""
      id="sales"
      type="Active"
    >
      <span
        class="Text-text-0 SideNavigationListItem-text-0"
      >
        Sales
      </span>
      <ui5-icon
        class="SideNavigationListItem-expandArrow-0"
        name="navigation-down-arrow"
      />
    </ui5-li-custom>
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0 SideNavigationListItem-noIcons-0"
      data-id="sales-opportunities"
      data-is-child=""
      id="sales-opportunities"
      style="--ui5-listitem-border-bottom: none;"
      type="Active"
    >
      <span
        class="Text-text-0 SideNavigationListItem-text-0"
      >
        My Opportunities
      </span>
    </ui5-li-custom>
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0 SideNavigationListItem-noIcons-0"
      data-id="sales-leads"
      data-is-child=""
      id="sales-leads"
      selected="true"
      style="--ui5-listitem-border-bottom: none;"
      type="Active"
    >
      <span
        class="Text-text-0 SideNavigationListItem-text-0"
      >
        My Leads
      </span>
    </ui5-li-custom>
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0 SideNavigationListItem-noIcons-0"
      data-id="sales-cpqs"
      data-is-child=""
      id="sales-cpqs"
      type="Active"
    >
      <span
        class="Text-text-0 SideNavigationListItem-text-0"
      >
        My CPQS
      </span>
    </ui5-li-custom>
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0 SideNavigationListItem-noIcons-0"
      data-id="delivieries"
      id="delivieries"
      type="Active"
    >
      <span
        class="Text-text-0 SideNavigationListItem-text-0"
      >
        Deliveries
      </span>
    </ui5-li-custom>
  </ui5-list>
  <span
    style="flex-grow: 1;"
  />
  <span
    class="SideNavigation-footerItemsSeparator-0"
  />
  <ui5-list
    mode="None"
    separators="All"
  >
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0 SideNavigationListItem-noIcons-0"
      data-id="1"
      id="1"
      type="Active"
    >
      <span
        class="Text-text-0 SideNavigationListItem-text-0"
      >
        Legal Information
      </span>
    </ui5-li-custom>
    <ui5-li-custom
      class="SideNavigationListItem-listItem-0 SideNavigationListItem-noIcons-0"
      data-id="2"
      id="2"
      type="Active"
    >
      <span
        class="Text-text-0 SideNavigationListItem-text-0"
      >
        Useful Links
      </span>
    </ui5-li-custom>
  </ui5-list>
</div>
`;
